*: Remove RocksDB

itsbilal opened the pull request:
As in title. Removes RocksDB and snappy, from the `c-deps` directory,
and updates tests and roachtests to reflect a Pebble-only world. Libroach
is slimmed to just the one method (`DumpThreadStacks`) that is still necessary.

As this is a massive change, I've tried to make it as easy to review as possible by
splitting changes in different commits. The first commit just updates temp file locking,
the second commit deletes RocksDB-specific code paths and updates the few remaining
RocksDB-only tests to use equivalent Pebble code paths, the fourth commit has
Makefile changes, and the last commit is a plain file deletion in `c-deps`. 

This is ready for review, though the following relatively-minor tasks are still TODO:

- [x] Resolve merge conflicts
- [x] Pass lint
- [x] Bring back thread stack dumps as they may be necessary for inspecting other CGo code

Commit:
storage: Use vfs.FS.Lock for temp file locking instead of RocksDB

Previously, we used RocksDB's file locking mechanism for temp directory
locking. As RocksDB is being removed from the codebase, move to using
Pebble's syscall-based mechanism.

Release note: None.

Commit:
*: Stop using RocksDB in Go code

This change deletes all RocksDB specific Go code, and updates
all tests that called RocksDB-specific methods to call their Pebble
versions instead.

This is a prerequisite for deleting RocksDB from the codebase.

Release note (general change): RocksDB can no longer be used as
the storage engine. Passing in --storage-engine=rocksdb will
return an error.

Commit:
cli,cmd: Update --storage-engine text to reflect RocksDB deletion

As --storage-engine=rocksdb no longer works as a command line
argument, update the help text for that flag to reflect its nature.

Also skip engine/switch, and inconsistency roachtests as they use
RocksDB-only functionality.

Release note (cli change): Update --storage-engine help text to
reflext RocksDB deletion.

Commit:
*: Update makefile to not build rocksdb, snappy

This change updates the Makefile to not build snappy, and rocksdb.
Libroach is built without those two libraries as well.

Release note: None.

Commit:
*: Delete RocksDB, snappy from c-deps

This change is a simple file deletion of:
c-deps/{rocksdb,snappy}{,-rebuild}

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/55509)
<!-- Reviewable:end -->

otan commented on pkg/cmd/roachtest/engine_switch.go:
```diff
@@ -140,6 +140,7 @@ func registerEngineSwitch(r *testRegistry) {
 	r.Add(testSpec{
 		Name:       fmt.Sprintf("engine/switch/nodes=%d", n),
 		Owner:      OwnerStorage,
+		Skip:       "rocksdb removed in 21.1",
```
since roachtest is always based on master, doesn't this block all invocations?

irfansharif replied:
It does. I'm told @jlinder+@adityamaru are working to address https://github.com/cockroachdb/cockroach/issues/51897 shortly.

knz commented:
jemalloc is needed for all the cgo code

irfansharif commented:
We'll need to do something here as well:

https://github.com/cockroachdb/cockroach/blob/69f0fb53c2ade2102713e686a7da68a726d8088c/pkg/cmd/roachtest/versionupgrade.go#L507-L508

irfansharif commented on pkg/cli/cliflags/flags.go:
```diff
@@ -827,8 +827,8 @@ Also, if you use equal signs in the file path to a store, you must use the
 		Description: `
-Storage engine to use for all stores on this cockroach node. Options are pebble
-or rocksdb. If unspecified, pebble is used.
+Storage engine to use for all stores on this cockroach node. The only option is pebble. Deprecated;
```
Would it make sense to remove the flag entirely?

itsbilal replied:
I've kept it around for now just so anyone forcibly using `--storage-engine=pebble` doesn't see it fail rightaway. But it's probably a good idea to remove it entirely.

otan replied:
probably makes sense to keep it for one more version, so that people migrating don't have to change flags in between right?

petermattis replied:
Agreed on keeping the flag for the 21.1 release. We may even want to support `--storage-engine=rocksdb` with a loud warning that you're actually using Pebble instead.

irfansharif commented on pkg/cli/debug.go:
```diff
@@ -646,24 +628,6 @@ func runDebugGCCmd(cmd *cobra.Command, args []string) error {
 	return nil
 }
 
-var debugRocksDBCmd = &cobra.Command{
```
I assume going forward the easiest way to use these subcommands (including the one I linked elsewhere for versionupgrade) is to use a 20.2 binary?

itsbilal replied:
Most of these subcommands are implemented (in a better way) under `cockroach debug pebble`. The few exceptions that don't exist are `checkpoint`, `put`, etc - and so I've just skipped the roachtests that use those until we've figured out an alternative way, or reimplemented them under pebble too. The underlying functionality exists in Pebble already; it's just not exposed to the command line.

itsbilal commented:
This is ready for a look. I could split the first 3 commits into a separate PR if that makes them easier to review, but the last two are fairly mechanical anyway so I don't expect much code reviewing discussion on those.

Only big piece that's still TODO is that thread stack dumps are disabled now as they are a casualty of the `libroach` deletion. If they're still worthwhile to keep, I could bring back just that part of libroach, or see if there's a way we can get a view of threads from Go. Thoughts welcome.

otan commented:
> Only big piece that's still TODO is that thread stack dumps are disabled now as they are a casualty of the libroach deletion. If they're still worthwhile to keep, I could bring back just that part of libroach, or see if there's a way we can get a view of threads from Go. Thoughts welcome.

fwiw, libroach was not hard to bazel-ify so i'm not opposed to keeping that part of libroach.

irfansharif commented on pkg/cmd/roachtest/versionupgrade.go:
```diff
@@ -503,7 +503,9 @@ func makeVersionFixtureAndFatal(
-			// without rocksdb (`make buildshort`).
+			// without rocksdb (`make buildshort`). Pebble already supports
+			// checkpointing (see CreateCheckpoint in pebble.go). This just needs
+			// to be exposed to a command line method.
```
I think we'll need to do it for this one before merge. Every release requires us to generate fixtures for it, and we do it through this codepath.

petermattis replied:
Ack. Probably worthwhile to add a `pebble checkpoint` tool. As Bilal mentioned, all the hard work is already done.

itsbilal replied:
This is resolved now that #55751 is in. I've also un-skipped the roachtest.

knz commented:
> fwiw, libroach was not hard to bazel-ify so i'm not opposed to keeping that part of libroach.

As long as we still have cgo code (and we do), it seems to be a good thing to keep?

Commit:
libroach: Slim libroach to only have the DBDumpThreadStacks method

This change deletes all of Libroach except for the methods
necessary to implement DBDumpThreadStacks. Also moves the Go-side
code to pkg/storage/stacks.go.

Release note: None.

itsbilal commented:
Sorry for the noise! But this is ready for another look. I've brought libroach back, just for thread stack dumping. Also fixed many tests, and working my way through the last minor build/CI-related issues. `version-upgrade` fixtures generation will be fixed when https://github.com/cockroachdb/pebble/pull/962 lands, which should happen before this lands.

itsbilal commented:
Latest push should result in a green build 🤞 . This is ready for a look now. Due to the large surface area of this PR, I have to rebase pretty regularly to prevent merge conflicts.

otan commented:
(LGTM, letting storage do sign off though)

sumeerbhola reviewed (approved):
I did not review the Makefile changes.
Rest looks great!

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 1 of 2 files at r1, 4 of 201 files at r6, 4 of 203 files at r11, 79 of 79 files at r12, 4 of 4 files at r13, 2 of 2 files at r14, 5 of 5 files at r15, 10 of 119 files at r16.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55509#-:-MK7REx-7uTn--XpcrlT:bd6tu0u)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @irfansharif, @itsbilal, @jbowens, @jlinder, @otan, @pbardea, and @petermattis)

___
*[pkg/storage/batch_test.go, line 804 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/55509#-MK7546i5r5WQkjSyAvv:-MK7546i5r5WQkjSyAvw:b7xsxvt) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc0bb2ed6b974283563213ac852cfa8c280e2f64/pkg/storage/batch_test.go#L804)):*
> ```Go
> }
> 
> func TestBatchBuilder(t *testing.T) {
> ```

are there existing corresponding tests that use `pebbleBatch`?
___
*[pkg/storage/batch_test.go, line 839 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/55509#-MK75GnG5jIEj5fGFTcW:-MK75GnG5jIEj5fGFTcX:b1gzt0j) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc0bb2ed6b974283563213ac852cfa8c280e2f64/pkg/storage/batch_test.go#L839)):*
> ```Go
> 			t.Fatal(err)
> 		}
> 		// TODO(itsbilal): Uncomment this when pebble.Batch supports SingleDeletion.
> ```

these old TODOs seem odd given that this test is using `rocksDBBatch`.
___
*[pkg/storage/bench_pebble_test.go, line 349 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/55509#-MK75vNL4xqjPtZH1wYc:-MK75vNL4xqjPtZH1wYd:b-epvc27) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9d115aabdbfb733a3dbaf685ea1f5f929b87f61/pkg/storage/bench_pebble_test.go#L349)):*
> ```Go
> }
> 
> func BenchmarkBatchBuilderPut(b *testing.B) {
> ```

Looks like this is the only thing we saved from `bench_rocksdb_test.go`. I guess the others were already mirrored above?
___
*[pkg/storage/engine_test.go, line 438 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/55509#-MK7LvrY2URFCosaY7nr:-MK7LvrY2URFCosaY7ns:bpmh5n6) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9d115aabdbfb733a3dbaf685ea1f5f929b87f61/pkg/storage/engine_test.go#L438)):*
> ```Go
> 	return mustMarshal(&v)
> }
> ```

I'm assuming that these are unchanged when copied from `rocksdb_merge_test.go`.
___
*[pkg/storage/sst_info_test.go, line 76 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/55509#-MK7NzbP-ayKP6M0NIfN:-MK7NzbP-ayKP6M0NIfO:bwpi36g) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9d115aabdbfb733a3dbaf685ea1f5f929b87f61/pkg/storage/sst_info_test.go#L76)):*
> ```Go
> }
> 
> func key(s string) MVCCKey {
> ```

Could this be more verbose, like `stringToKeyForTesting`?
___
*[pkg/storage/temp_dir_windows.go, line 19 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/55509#-MK6zHDV5aWG6e7BXY5j:-MK6zHDWDwUcvNThBE32:b-jk5cvc) ([raw file](https://github.com/cockroachdb/cockroach/blob/a21cd8b8acb318cc71bc9fd4b2f621538984bfcc/pkg/storage/temp_dir_windows.go#L19)):*
> ```Go
> // lockFile sets a lock on the specified file. No-op.
> func lockFile(filename string) (lockStruct, error) {
> 	return lockStruct{}, nil
> ```

were these always noop? Can you add a comment stating why this is ok?


<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
Just saw one of Sumeer's comments that deserved a response. Happened to notice that there are some `*.rej` files in the PR which should be removed.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55509#-:-MK7fbQn1vY6Fytf-EW_:b-byg5pp)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @irfansharif, @itsbilal, @jbowens, @jlinder, @otan, @pbardea, @petermattis, and @sumeerbhola)

___
*[pkg/storage/temp_dir_unix.go, line 29 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/55509#-MK7fTbWCLKA1VJeA0l6:-MK7fTbWCLKA1VJeA0l7:bduoos7) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ae81bfe6dab3238f0822a6f3c31d2fc85d160f7/pkg/storage/temp_dir_unix.go#L29)):*
> ```Go
> // lockFile sets a lock on the specified file, using flock.
> func lockFile(filename string) (lockStruct, error) {
> 	file, err := os.OpenFile(filename, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0644)
> ```

We should be using `pebble/vfs.FS.Lock()` here. That will also work on Windows.
___
*[pkg/storage/temp_dir_windows.go, line 19 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/55509#-MK6zHDV5aWG6e7BXY5j:-MK7ewbT4TdEMV7zHodO:b-5ys8q7) ([raw file](https://github.com/cockroachdb/cockroach/blob/a21cd8b8acb318cc71bc9fd4b2f621538984bfcc/pkg/storage/temp_dir_windows.go#L19)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

were these always noop? Can you add a comment stating why this is ok?
</blockquote></details>

I think we should be using `pebble/vfs.FS.Lock()` here.


<!-- Sent from Reviewable.io -->

itsbilal reviewed (commented):
Thanks for the reviews! Made the suggested changes.

I didn't see any *.rej files remaining in any of the commits. Reviewable shows some as they existed in an old revision, but I deleted them after and amended the commits.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55509#-:-MKAbKKMAGYN37_hvmLy:b4i7qrs)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @irfansharif, @jbowens, @jlinder, @otan, @pbardea, @petermattis, and @sumeerbhola)

___
*[pkg/storage/batch_test.go, line 804 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/55509#-MK7546i5r5WQkjSyAvv:-MKAZqRz5U_vjDAIroqW:b-wriy7q) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc0bb2ed6b974283563213ac852cfa8c280e2f64/pkg/storage/batch_test.go#L804)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

are there existing corresponding tests that use `pebbleBatch`?
</blockquote></details>

Not really. The test was basically one to check for Repr equivalence between doing `db*` operations on a RocksDB batch, and using the previously-separate RocksDBBatchBuilder (which, under the hood, has been using Pebble.Batch for a couple months). Since there's really no Pebble equivalent of the former, checking the latter against itself would be pretty pointless, so I deleted the test entirely. 
___
*[pkg/storage/batch_test.go, line 839 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/55509#-MK75GnG5jIEj5fGFTcW:-MKA_GtW9jDUc5pvaUBZ:b-euqzlc) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc0bb2ed6b974283563213ac852cfa8c280e2f64/pkg/storage/batch_test.go#L839)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

these old TODOs seem odd given that this test is using `rocksDBBatch`.
</blockquote></details>

It was comparing rocksDBBatch against Pebble.Batch, so I guess it did make sense back then. 
___
*[pkg/storage/bench_pebble_test.go, line 349 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/55509#-MK75vNL4xqjPtZH1wYc:-MKA_QuZ7RCT-Cd-lSET:b-sydryd) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9d115aabdbfb733a3dbaf685ea1f5f929b87f61/pkg/storage/bench_pebble_test.go#L349)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Looks like this is the only thing we saved from `bench_rocksdb_test.go`. I guess the others were already mirrored above?
</blockquote></details>

Yes. The rest are mirrored.
___
*[pkg/storage/engine_test.go, line 438 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/55509#-MK7LvrY2URFCosaY7nr:-MKAaVirB09T67MVbnQr:b-42xi5x) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9d115aabdbfb733a3dbaf685ea1f5f929b87f61/pkg/storage/engine_test.go#L438)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I'm assuming that these are unchanged when copied from `rocksdb_merge_test.go`.
</blockquote></details>

Yes, they are just the couple test helpers that are still being used. Copied verbatim.
___
*[pkg/storage/sst_info_test.go, line 76 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/55509#-MK7NzbP-ayKP6M0NIfN:-MKAay7x9Iown5y3dfs3:b-3036rj) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9d115aabdbfb733a3dbaf685ea1f5f929b87f61/pkg/storage/sst_info_test.go#L76)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Could this be more verbose, like `stringToKeyForTesting`?
</blockquote></details>

Done. Renamed it to `stringToKey`; the ForTesting part is implied since it's in a _test file. That said, it was also a verbatim copy.
___
*[pkg/storage/temp_dir_unix.go, line 29 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/55509#-MK7fTbWCLKA1VJeA0l6:-MKAbEoUAKJMxGmxihqe:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ae81bfe6dab3238f0822a6f3c31d2fc85d160f7/pkg/storage/temp_dir_unix.go#L29)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

We should be using `pebble/vfs.FS.Lock()` here. That will also work on Windows.
</blockquote></details>

Done.
___
*[pkg/storage/temp_dir_windows.go, line 19 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/55509#-MK6zHDV5aWG6e7BXY5j:-MKAbE_08shWKQxtO__L:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a21cd8b8acb318cc71bc9fd4b2f621538984bfcc/pkg/storage/temp_dir_windows.go#L19)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I think we should be using `pebble/vfs.FS.Lock()` here.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
> I didn't see any *.rej files remaining in any of the commits. Reviewable shows some as they existed in an old revision, but I deleted them after and amended the commits.

Ah, perhaps they were deleted and I didn't notice that. Carry on.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55509#-:-MKBdU6g6nMGjxgxHA2F:bdlhivw)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @irfansharif, @itsbilal, @jbowens, @jlinder, @otan, @pbardea, and @sumeerbhola)

___
*[.gitmodules, line 7 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/55509#-MKBde1681PPv_7UPRs-:-MKBde1681PPv_7UPRs0:br3fxq6) ([raw file](https://github.com/cockroachdb/cockroach/blob/a66a97104f52e79295681c55963913ffb2c596bc/.gitmodules#L7)):*
> ```gitconfig
> 	path = c-deps/jemalloc
> 	url = https://github.com/cockroachdb/jemalloc.git
> [submodule "c-deps/cryptopp"]
> ```

Is cryptopp still needed? I thought it was only used for RocksDB encryption-at-rest.
___
*[pkg/cmd/roachtest/inconsistency.go, line 25 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/55509#-MKBeWuN84hfhFyLgMSQ:-MKBeWuN84hfhFyLgMSR:bumb60n) ([raw file](https://github.com/cockroachdb/cockroach/blob/a66a97104f52e79295681c55963913ffb2c596bc/pkg/cmd/roachtest/inconsistency.go#L25)):*
> ```Go
> 		Name:       fmt.Sprintf("inconsistency"),
> 		Owner:      OwnerKV,
> 		Skip:       "Uses RocksDB put command; unskip when that's bypassed",
> ```

Is the intention to fix this soon? I wouldn't want to leave this test skipped for too long. Perhaps file an issue.
___
*[pkg/kv/kvserver/store.go, line 855 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/55509#-MKBernxEZdDnelpc25I:-MKBernxEZdDnelpc25J:b5otcv) ([raw file](https://github.com/cockroachdb/cockroach/blob/a66a97104f52e79295681c55963913ffb2c596bc/pkg/kv/kvserver/store.go#L855)):*
> ```Go
> 	//
> 	// TODO(bilal): Delete this and all compactor-related code, now that Pebble is
> 	// the only storage engine.
> ```

File an issue unless you plan to do this immediately.
___
*[pkg/server/debug/server.go, line 183 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/55509#-MKBf5R6DHtT9FNBChff:-MKBf5R6DHtT9FNBChfg:b-9cichm) ([raw file](https://github.com/cockroachdb/cockroach/blob/d775b9819626ca03f1fa8682abc1709fa16c99fb/pkg/server/debug/server.go#L183)):*
> ```Go
> 	})
> 
> 	mux.HandleFunc("/debug/threads", func(w http.ResponseWriter, req *http.Request) {
> ```

Should this be left in? I think we want to keep the thread-stack dumps.
___
*[pkg/storage/pebble_iterator.go, line 298 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/55509#-MKBfesvAU76KvF_BGQn:-MKBfesw9fQ12ZS317HA:b-zg48sh) ([raw file](https://github.com/cockroachdb/cockroach/blob/a66a97104f52e79295681c55963913ffb2c596bc/pkg/storage/pebble_iterator.go#L298)):*
> ```Go
> // NoSplitSpans.
> func isValidSplitKey(key roachpb.Key, noSplitSpans []roachpb.Span) bool {
> 	if key.Equal(keys.Meta2KeyMax) {
> ```

Is this a real bug fix? Does it need to be backported to 20.2 and 20.1?
___
*[pkg/storage/enginepb/engine.proto, line 25 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/55509#-MKBfxpt2ZITlYqKVIso:-MKBfxpt2ZITlYqKVIsp:b-kczoka) ([raw file](https://github.com/cockroachdb/cockroach/blob/d775b9819626ca03f1fa8682abc1709fa16c99fb/pkg/storage/enginepb/engine.proto#L25)):*
> ```protobuf
>   EngineTypeDefault = 0;
>   // Denotes RocksDB as the underlying storage engine type.
>   EngineTypeRocksDB = 1;
> ```

Is it valid to use `reserved = 1` in an enum?


<!-- Sent from Reviewable.io -->

itsbilal reviewed (commented):
TFTRs! Again, sorry for the force-push noise; lots of subtle build issues that are only happening on TC but not locally. But most of them should be behind us now.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55509#-:-MKBsaUc1_beVJFtFlOu:b-o9lxcr)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @irfansharif, @jbowens, @jlinder, @otan, @pbardea, @petermattis, and @sumeerbhola)

___
*[.gitmodules, line 7 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/55509#-MKBde1681PPv_7UPRs-:-MKBqoW4AIa6mssuONbs:bui4ggs) ([raw file](https://github.com/cockroachdb/cockroach/blob/a66a97104f52e79295681c55963913ffb2c596bc/.gitmodules#L7)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Is cryptopp still needed? I thought it was only used for RocksDB encryption-at-rest.
</blockquote></details>

It's not. @otan pointed this out too. I'll fix this right away in a follow-up PR as this one is already large as-is.
___
*[pkg/cmd/roachtest/inconsistency.go, line 25 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/55509#-MKBeWuN84hfhFyLgMSQ:-MKBqu8P12K4dRH83GQi:b-ldtnv1) ([raw file](https://github.com/cockroachdb/cockroach/blob/a66a97104f52e79295681c55963913ffb2c596bc/pkg/cmd/roachtest/inconsistency.go#L25)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Is the intention to fix this soon? I wouldn't want to leave this test skipped for too long. Perhaps file an issue.
</blockquote></details>

I'll file an issue, thanks for the reminder!
___
*[pkg/kv/kvserver/store.go, line 855 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/55509#-MKBernxEZdDnelpc25I:-MKBqxVW8knZMnOMPnmC:b4a2ls6) ([raw file](https://github.com/cockroachdb/cockroach/blob/a66a97104f52e79295681c55963913ffb2c596bc/pkg/kv/kvserver/store.go#L855)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

File an issue unless you plan to do this immediately.
</blockquote></details>

Filing an issue.
___
*[pkg/server/debug/server.go, line 183 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/55509#-MKBf5R6DHtT9FNBChff:-MKBqz506DrBaD17R6up:b-x2vft9) ([raw file](https://github.com/cockroachdb/cockroach/blob/d775b9819626ca03f1fa8682abc1709fa16c99fb/pkg/server/debug/server.go#L183)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Should this be left in? I think we want to keep the thread-stack dumps.
</blockquote></details>

Done. Added it back; deleted it by mistake.
___
*[pkg/storage/pebble_iterator.go, line 298 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/55509#-MKBfesvAU76KvF_BGQn:-MKBr0Qm2qg5_K6rh92x:b-b18ao6) ([raw file](https://github.com/cockroachdb/cockroach/blob/a66a97104f52e79295681c55963913ffb2c596bc/pkg/storage/pebble_iterator.go#L298)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Is this a real bug fix? Does it need to be backported to 20.2 and 20.1?
</blockquote></details>

It's not an issue as the old code is a bit more strict (i.e. it would advise against splitting at start keys of all spans, new code allows splitting at all start keys except for the Meta2 one). Might be worth getting a second opinion from KV ( @irfansharif ?), but basically the new disallowed-set is a subset of the old one, so we're actually less strict now.
___
*[pkg/storage/enginepb/engine.proto, line 25 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/55509#-MKBfxpt2ZITlYqKVIso:-MKBrqCIAFpAGVNY16wJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d775b9819626ca03f1fa8682abc1709fa16c99fb/pkg/storage/enginepb/engine.proto#L25)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Is it valid to use `reserved = 1` in an enum?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

petermattis reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> modulo questions about the `isValidSplitKey` change. I think that change is good, just wondering if it should be pulled into a separate PR.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55509#-:-MKBvcNKEEbUSYL8dhIb:b-8drtxd)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @irfansharif, @jbowens, @jlinder, @otan, @pbardea, and @sumeerbhola)

___
*[.gitmodules, line 7 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/55509#-MKBde1681PPv_7UPRs-:-MKBtVDI311IcRLXsMiz:b6shgiq) ([raw file](https://github.com/cockroachdb/cockroach/blob/a66a97104f52e79295681c55963913ffb2c596bc/.gitmodules#L7)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

It's not. @otan pointed this out too. I'll fix this right away in a follow-up PR as this one is already large as-is.
</blockquote></details>

Ack. :thumbsup: 
___
*[pkg/storage/pebble_iterator.go, line 298 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/55509#-MKBfesvAU76KvF_BGQn:-MKBuVc-BHxqj97hQvBC:b-lstiob) ([raw file](https://github.com/cockroachdb/cockroach/blob/a66a97104f52e79295681c55963913ffb2c596bc/pkg/storage/pebble_iterator.go#L298)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

It's not an issue as the old code is a bit more strict (i.e. it would advise against splitting at start keys of all spans, new code allows splitting at all start keys except for the Meta2 one). Might be worth getting a second opinion from KV ( @irfansharif ?), but basically the new disallowed-set is a subset of the old one, so we're actually less strict now.
</blockquote></details>

Ack. I'd like a second opinion. This feels like something that could get pulled out into a separate commit/PR for easy backporting. I agree that the change to use `ProperlyContainsKey` is a loosening of the split point logic, but the disallowing of splits at `Meta2KeyMax` is a strengthening of that logic. 


<!-- Sent from Reviewable.io -->

itsbilal reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55509)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @irfansharif, @jbowens, @jlinder, @otan, @pbardea, and @sumeerbhola)

___
*[pkg/storage/pebble_iterator.go, line 298 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/55509#-MKBfesvAU76KvF_BGQn:-MKBwU3OEOYx2bXPQIOl:b-mngz4l) ([raw file](https://github.com/cockroachdb/cockroach/blob/a66a97104f52e79295681c55963913ffb2c596bc/pkg/storage/pebble_iterator.go#L298)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Ack. I'd like a second opinion. This feels like something that could get pulled out into a separate commit/PR for easy backporting. I agree that the change to use `ProperlyContainsKey` is a loosening of the split point logic, but the disallowing of splits at `Meta2KeyMax` is a strengthening of that logic. 
</blockquote></details>

Yep, and the key point is that `Meta2KeyMax` is already one of the start keys in noSplitSpans, so it would have been disallowed in both the old code (through `ContainsKey`), and in the new code (through the special case). The loosening is effectively only with all the other disallowed spans.


<!-- Sent from Reviewable.io -->

petermattis reviewed (approved):
:shipit: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55509#-:-MKFCset-R-7UCorDmol:bizlbj6)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @irfansharif, @jbowens, @jlinder, @otan, @pbardea, and @sumeerbhola)

___
*[pkg/storage/pebble_iterator.go, line 298 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/55509#-MKBfesvAU76KvF_BGQn:-MKFCdXgEtCQKRRGiB-T:b-b0uhs1) ([raw file](https://github.com/cockroachdb/cockroach/blob/a66a97104f52e79295681c55963913ffb2c596bc/pkg/storage/pebble_iterator.go#L298)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

Yep, and the key point is that `Meta2KeyMax` is already one of the start keys in noSplitSpans, so it would have been disallowed in both the old code (through `ContainsKey`), and in the new code (through the special case). The loosening is effectively only with all the other disallowed spans.
</blockquote></details>

Ah, I was missing that `Meta2KeyMax` was a start key of one of the spans in `noSplitSpans`. So we were always disabling splits at `Meta2KeyMax`, but with the change to use `ProperlyContainsKey` below we wouldn't be. 


<!-- Sent from Reviewable.io -->

itsbilal commented:
Time to roll with it then. Thanks for all the reviews!

bors r+
